src/net/dns.cc

```diff
@@ -845,21 +846,31 @@ class net::dns_resolver::impl
                     }).finally([fd, me = shared_from_this()] {
                         me->release(fd);
                     });
-                    // if we have a fast-fail, give error.
-                    if (e.udp.f.failed()) {
-                        try {
-                            e.udp.f.get();
-                        } catch (std::system_error& e) {
-                            errno = e.code().value();
-                        } catch (...) {
+
+                    if (e.upd.f.available()) {
+                        // if we have a fast-fail, give error.
+                        if (!e.udp.f.failed()) {
```

Comment:
Code becomes shorter (and imho more readable) if you make it:

```
                    if (e.upd.f.available()) {
                        // if we have a fast-fail, give error.
                        if (e.udp.f.failed()) {
                            try {
                                e.udp.f.get();
                            } catch (std::system_error& e) {
                                errno = e.code().value();
                            } catch (...) {
                            }
                            e.udp.f = make_ready_future<>();
                            return -1;
                        }
                    } else {
                        // ensure that no exception from channel.send is left uncaught
                        e.udp.f = e.udp.f.handle_exception_type([](std::system_error const& e){
                            dns_log.warn("UDP send exception: {}", e.what());
                        });
                    }
                    // c-ares does _not_ use non-blocking retry for udp sockets. We just pretend
                    // all is fine even though we have no idea. Barring stack/adapter failure it
                    // is close to the same guarantee a "normal" message send would have anyway.
                    return bytes;
                default:
                    return -1;

Reply:
agreed, done